auslib/test/blobs/test_apprelease.py

```diff
@@ -2847,6 +2849,412 @@ def testSchema8OptionalAttributes(self):
+        {"versions": [">=49.0"]}
+    ),
+    (
+        {"locales": ["ja"]},
+        {"locales": ["ja-JP-mac"]},
+    ),
+    (
+        {"locales": ["de"], "channels": ["beta*"], "versions": ["<50.0"]},
+        {"locales": ["de", "fr"], "channels": ["release"], "versions": ["49.0"]}
+    ),
+    (
+        {"locales": ["de"], "channels": ["release"], "versions": ["<50.0"]},
+        {"locales": ["en-US", "fr"], "channels": ["release"], "versions": ["49.0"]}
+    ),
+    (
+        {"locales": ["de"], "channels": ["release"], "versions": ["<48.0"]},
+        {"locales": ["de", "fr"], "channels": ["release"], "versions": ["49.0"]}
+    ),
+])
+def testSchema9CanCreateValidBlobs(for1, for2):
+    good_blob = {
+        "name": "bad",
+        "schema_version": 9,
+        "hashFunction": "sha512",
+        "appVersion": "31.0.2",
+        "displayVersion": "31.0.2",
+        "updateLine": [
+            {
+                "for": for1,
+                "fields": {
+                    "detailsURL": "http://example.org/details/%LOCALE%",
+                    "type": "minor"
+                }
+            },
+            {
+                "for": for2,
+                "fields": {
+                    "detailsURL": "http://example.org/specialdetails/%LOCALE%",
+                }
+            }
+        ]
+    }
+    blob = ReleaseBlobV9(**good_blob)
+    blob.validate("h", {'a.com': ('h',)})
+
+
+@pytest.mark.parametrize("for1,for2", [
+    (
+        {},
+        {},
+    ),
+    (
+        {},
+        {"locales": ["de", "fr"]}
+    ),
+    (
+        {"locales": ["de"]},
+        {"locales": ["de", "fr"]}
+    ),
+    (
+        {},
+        {"channels": ["release"]}
+    ),
+    (
+        {"channels": ["release"]},
+        {"channels": ["release-cck-foo"]}
+    ),
+    (
+        {"channels": ["release*"]},
+        {"channels": ["release-cck-foo"]}
+    ),
+    (
+        {"channels": ["release-cck-foo"]},
+        {"channels": ["release-cck-*"]}
+    ),
+    (
+        {"channels": ["release*"]},
+        {"channels": ["release-cck-*"]}
+    ),
+    (
+        {"channels": ["release"]},
+        {"channels": ["release*"]}
+    ),
+    (
+        {},
+        {"versions": ["50.0"]}
+    ),
+    (
+        {"versions": ["<50.0"]},
+        {"versions": ["49.0"]}
+    ),
+    (
+        {"versions": ["49.0"]},
+        {"versions": ["<=49.0"]}
+    ),
+    (
+        {"versions": ["<50.0"]},
+        {"versions": ["<49.0"]}
+    ),
+    (
+        {"versions": [">50.0"]},
+        {"versions": [">49.0"]}
+    ),
+    (
+        {"versions": ["<50.0"]},
+        {"versions": [">49.0"]}
+    ),
+    (
+        {"versions": ["<50.0"]},
+        {"versions": [">=49.0"]}
+    ),
+    (
+        {"versions": [">49.0"]},
+        {"versions": ["<=50.0"]}
+    ),
+    (
+        {"versions": [">=49.0"]},
+        {"versions": ["<=50.0"]}
+    ),
+    (
+        {"versions": ["<=49.0"]},
+        {"versions": ["<=50.0"]}
```

Comment:
I was playing around in this space because `decrement_version()` seems like a good place for unexpected things to happen. For example, if the human is sloppy and puts `<49` instead of `<49.0` then it's decremented to 48 instead of 48.99, but in practice `MozillaVersion()` will raise `BadDataError: Version number 49 is invalid` before any comparison. I'm not sure if the there is anything to prevent malformed input but it could be quite useful.

The code performed nicely with tests like `<49.0` with `<48.0.4`, `<49.0.2` with `<49.0`, and `<49.0.1` with `<49.0`.

Reply:
Good find here. Even though we can't find a way to decrement badly right now, it doesn't mean it can't happen in the future (eg: if MozillaVersion ever gets replacde/removed).

What if we tweak increment/decrement to add a .0 if there's only one part to the version? Considering these are only meant to be used with version strings (not numbers in general) it seems like a reasonable safeguard.

Reply:
I think it would also be OK to insist that version conditions use well formed versions when submitting a blob, and reject any that don't - less magic behind the scenes then. Either way, lets make sure rules and WNP require the same thing. Was looking at the docs to see if I could figure that out and realized we'll need some doc changes for this work.

Reply:
On IRC, I believe we said that we want to make sure that Rules can't have versions like "4" or "<4" (single part version numbers) to avoid differences between its matching and what we do for updateLine. Turns out we already do that, via the swagger spec! https://screenshots.firefox.com/R7vxOpgKb3Lv34K5/localhost

`version` has a format defined in https://github.com/mozilla/balrog/blob/master/auslib/web/common/swagger/definitions.yml#L222, which is implemented in https://github.com/mozilla/balrog/blob/master/auslib/web/admin/views/validators.py#L58, which attempts to parse incoming versions with MozillaVersion.

I'm adding tests to make sure this doesn't regress, and I'll update the docs.

Reply:
Ooh, that's cool! Can we leverage version_validator() by using `type: version` in apprelease-v9.yml at line 69 ? ie blobs are checked on the way into the app.

Reply:
I think they're already checked because .validate() is called by the database layer before insert/update - but it would be good to have the web layer do this (it would be faster, and we'd get a better error message). I'll look into it.

Reply:
...except that .validate() doesn't do any manual validation against MozillaVersion. I'm pushing an update that enables that via the jsonschema validator.